src/category_theory/monoidal/coherence.lean

```diff
@@ -0,0 +1,302 @@
+{ lift := (λ_ (lift_obj.lift X)).inv, }
+instance lift_hom_right_unitor_hom (X : C) [lift_obj X] : lift_hom (ρ_ X).hom :=
+{ lift := (ρ_ (lift_obj.lift X)).hom, }
+instance lift_hom_right_unitor_inv (X : C) [lift_obj X] : lift_hom (ρ_ X).inv :=
+{ lift := (ρ_ (lift_obj.lift X)).inv, }
+instance lift_hom_associator_hom (X Y Z : C) [lift_obj X] [lift_obj Y] [lift_obj Z] :
+  lift_hom (α_ X Y Z).hom :=
+{ lift := (α_ (lift_obj.lift X) (lift_obj.lift Y) (lift_obj.lift Z)).hom, }
+instance lift_hom_associator_inv (X Y Z : C) [lift_obj X] [lift_obj Y] [lift_obj Z] :
+  lift_hom (α_ X Y Z).inv :=
+{ lift := (α_ (lift_obj.lift X) (lift_obj.lift Y) (lift_obj.lift Z)).inv, }
+instance lift_hom_comp {X Y Z : C} [lift_obj X] [lift_obj Y] [lift_obj Z] (f : X ⟶ Y) (g : Y ⟶ Z)
+  [lift_hom f] [lift_hom g] : lift_hom (f ≫ g) :=
+{ lift := lift_hom.lift f ≫ lift_hom.lift g }
+instance lift_hom_tensor {W X Y Z : C} [lift_obj W] [lift_obj X] [lift_obj Y] [lift_obj Z]
+  (f : W ⟶ X) (g : Y ⟶ Z) [lift_hom f] [lift_hom g] : lift_hom (f ⊗ g) :=
+{ lift := lift_hom.lift f ⊗ lift_hom.lift g }
+
+/--
+A typeclass carrying a choice of monoidal structural isomorphism between two objects.
+Used by the `⊗≫` monoidal composition operator, and the `coherence` tactic.
+-/
+-- We could likely turn this into a `Prop` valued existential if that proves useful.
+class monoidal_coherence (X Y : C) [lift_obj X] [lift_obj Y] :=
+(hom [] : X ⟶ Y)
+[is_iso : is_iso hom . tactic.apply_instance]
+
+attribute [instance] monoidal_coherence.is_iso
+
+namespace monoidal_coherence
+
+@[simps]
+instance refl (X : C) [lift_obj X] : monoidal_coherence X X := ⟨𝟙 _⟩
+
+@[simps]
+instance tensor (X Y Z : C) [lift_obj X] [lift_obj Y] [lift_obj Z] [monoidal_coherence Y Z] :
+  monoidal_coherence (X ⊗ Y) (X ⊗ Z) :=
+⟨𝟙 X ⊗ monoidal_coherence.hom Y Z⟩
+
+@[simps]
+instance left (X Y : C) [lift_obj X] [lift_obj Y] [monoidal_coherence X Y] :
+  monoidal_coherence (𝟙_ C ⊗ X) Y :=
+⟨(λ_ X).hom ≫ monoidal_coherence.hom X Y⟩
+
+@[simps]
+instance left' (X Y : C) [lift_obj X] [lift_obj Y] [monoidal_coherence X Y] :
+  monoidal_coherence X (𝟙_ C ⊗ Y) :=
+⟨monoidal_coherence.hom X Y ≫ (λ_ Y).inv⟩
+
+@[simps]
+instance right (X Y : C) [lift_obj X] [lift_obj Y] [monoidal_coherence X Y] :
+  monoidal_coherence (X ⊗ 𝟙_ C) Y :=
+⟨(ρ_ X).hom ≫ monoidal_coherence.hom X Y⟩
+
+@[simps]
+instance right' (X Y : C) [lift_obj X] [lift_obj Y] [monoidal_coherence X Y] :
+  monoidal_coherence X (Y ⊗ 𝟙_ C) :=
+⟨monoidal_coherence.hom X Y ≫ (ρ_ Y).inv⟩
+
+@[simps]
+instance assoc (X Y Z W : C) [lift_obj W] [lift_obj X] [lift_obj Y] [lift_obj Z]
+  [monoidal_coherence (X ⊗ (Y ⊗ Z)) W] : monoidal_coherence ((X ⊗ Y) ⊗ Z) W :=
+⟨(α_ X Y Z).hom ≫ monoidal_coherence.hom (X ⊗ (Y ⊗ Z)) W⟩
+
+@[simps]
+instance assoc' (W X Y Z : C) [lift_obj W] [lift_obj X] [lift_obj Y] [lift_obj Z]
+  [monoidal_coherence W (X ⊗ (Y ⊗ Z))] : monoidal_coherence W ((X ⊗ Y) ⊗ Z) :=
+⟨monoidal_coherence.hom W (X ⊗ (Y ⊗ Z)) ≫ (α_ X Y Z).inv⟩
+
```

Comment:
It seems that `monoidal_coherence X ((X ⊗ 𝟙_ C) ⊗ 𝟙_ C)` succeeds by `apply_instance` but `monoidal_coherence X (X ⊗ (𝟙_ C ⊗ 𝟙_ C))` doesn't. Is there a systematic way to give an instance like this?

Reply:
Oooh, good catch, thank you. I added two instances `tensor_right` and `tensor_right'` that cover this situation.

Finding more problems, or an (informal) argument that these typeclasses suffice, would be extremely helpful! :-)